cluster/event_monitor.go

```diff
@@ -25,15 +27,36 @@ type decodingResult struct {
 // NewEventsMonitor returns an EventsMonitor
 func NewEventsMonitor(cli client.APIClient, handler func(msg events.Message) error) *EventsMonitor {
 	return &EventsMonitor{
-		cli:     cli,
-		handler: handler,
+		stopChan: make(chan struct{}),
+		errChan:  make(chan error),
+		cli:      cli,
+		handler:  handler,
 	}
 }
 
-// Start starts the EventsMonitor
-func (em *EventsMonitor) Start(ec chan error) {
-	em.stopChan = make(chan struct{})
+// Start starts a goroutine and calls function start to listen event stream.
+// In the goroutine, if errChan is not nil, then call Start() recursively.
+func (em *EventsMonitor) Start() {
+	go func() {
+		if err := <-em.errChan; err != nil {
+			// Swarm needs to wait an interval to start eventsMonitor.
+			// If the Engine is lost forever, before Swarm judges it as disconnected，
+			// Swarm will enter a loop to start eventsMonitor which may consume CPU a lot.
+			time.Sleep(1 * time.Second)
+
+			// EventMonitor ran into an error, start it again.
+			em.Start()
+		} else {
+			// if err is nil, it means stopChan is closed and EventMonitor should return as well.
+			return
+		}
+	}()
+
+	em.start(em.errChan)
+}
 
+// start starts to listen event stream of docker engine.
+func (em *EventsMonitor) start(ec chan error) {
```

Comment:
AFAIK, if we use `start` and `Start`, they are same function with minor difference and one call the other. Here is not the case. I suggest choosing better name.